### ISSUE TITLE
Add macOS ARM build (universal or additional binaries)

### ISSUE THREAD

**Issue by jonasschnelli** [bitcoin/bitcoin#20371]
First ARM based Macs are available now.
See the porting guide https://developer.apple.com/documentation/xcode/porting_your_macos_apps_to_apple_silicon and the architectural differences: https://developer.apple.com/documentation/apple_silicon/addressing_architectural_differences_in_your_macos_code

Things to do:
* [x] fix depends build on ARM with macOS 10.16 (done in merged #20482)
* [x] fix CRC32 build issue (fix in #20603)
* [x] update macOS build readme (if required)
* [x] allow cross compiling ARM64 Mac on linux/gitian
* [x] fix `make deploy` on ARM64

**Comment by decryp2kanon** [comment]
apple ARM is running aarch64 binary?

**Comment by jonasschnelli** [comment]
Support for AArch64 NI SHA2 instructions should/could also be added.

**Comment by jonasschnelli** [comment]
Building on a ARM Mac with M1 via the depends build systems works with a few tweaks (https://github.com/bitcoin/bitcoin/pull/20482 plus crc32 getauxval fix).

The harder part is now to get cross compilation to work.
It probably requires a newer clang then the just merged upgrade to 8 (we need `arm-apple-darwin`).
The Xcode 12.12 SDK is probably required ([here](https://bitcoin.jonasschnelli.ch/gitian/Xcode-12.2-12B45b-extracted-SDK-with-libcxx-headers.tar.gz)).
Qt 5.9 works for cross compilation but has visual glitches on macOS11.
Update to Qt LTS 5.15 might be a good idea (#19716).

**Comment by satindergrewal** [comment]
I got error at with `make` command in root of `bitcoin` directory after manually applying patch to `qt.mk`. These are the steps I followed:

Installed Homebrew with this script:
https://gist.github.com/nrubin29/bea5aa83e8dfa91370fe83b62dad6dfa

command:
```
/bin/bash -c "$(curl -fsSL https://gist.githubusercontent.com/nrubin29/bea5aa83e8dfa91370fe83b62dad6dfa/raw/48f48f7fef21abb308e129a80b3214c2538fc611/homebrew_m1.sh)"
source ~/.zshrc
```

#### bitcoin install steps
```
git clone https://github.com/bitcoin/bitcoin
cd bitcoin/

# manually edited depends/packages/qt.mk file as per this git PR:
# https://github.com/bitcoin/bitcoin/pull/20482/files

cd depends/
make
cd ..
./configure --prefix=$(pwd)/depends/arm-apple-darwin20.1.0  --with-gui=yes --disable-tests --disable-bench --enable-upnp-default --enable-experimental-asm --enable-static --disable-shared
make
```

got this error:
```
  CXX      leveldb/helpers/memenv/libmemenv_a-memenv.o
  AR       leveldb/libmemenv.a
gcc -I. -I./src -Wall -Wextra -Wno-unused-function -g -O2 -c src/gen_context.c -o gen_context.o
gcc -Wall -Wextra -Wno-unused-function -g -O2  gen_context.o -o gen_context
./gen_context
  CC       src/libsecp256k1_la-secp256k1.lo
  CCLD     libsecp256k1.la
  CXXLD    bitcoind
Undefined symbols for architecture arm64:
  "_getauxval", referenced from:
      crc32c::Extend(unsigned int, unsigned char const*, unsigned long) in libcrc32c.a(libcrc32c_a-crc32c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bitcoind] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

Is that expected? or did I miss anything?

**Comment by jonasschnelli** [comment]
@satindergrewal 
Yes. Somehow the crc32 library thinks we are on ARM linux and tries to use getauxval.

A quick fix is to change `getauxval(AT_HWCAP);` to `getauxval1(AT_HWCAP);` in `configure.ac` (I know a very lame fix!).
I'll do a proper PR soon.

**Comment by jonasschnelli** [comment]
Related #20603

**Comment by dongcarl** [comment]
I think we might have to bump to Qt 6.1 for this: https://bugreports.qt.io/browse/QTBUG-85279
There exists a qt5commpat layer which might be helpful for the transition: https://www.qt.io/blog/porting-from-qt-5-to-qt-6-using-qt5compat-library

Qt 6 also replaced qmake with cmake, which is much nicer to work with.

**Comment by jonasschnelli** [comment]
@dongcarl:
With the current master, building depends (as it is) works on an ARM64 Mac. Do you think we have to bump Qt to make cross compiling arm-apple-dawrin on a x86 linux host work?

I really have little understanding about cross compiling. Would appreciate if someone could look into building depends for `HOST=arm-apple-darwin20`

**Comment by awesome-doge** [comment]
I have referenced the above solution and successfully compiled bitcoin.
But I have a problem with the last command `make deploy`.

https://gist.github.com/awesome-doge/9bfbd1d3fd9b725c4df224433fa3596e#file-build-apple-m1-version-of-bitcoin-md

```
➜  bitcoin git:(master) make deploy

build-aux/install-sh -c -d Bitcoin-Qt.app/Contents
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents/Resources
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents/Resources
/usr/bin/install -c -m 644 src/qt/res/icons/bitcoin.icns Bitcoin-Qt.app/Contents/Resources/bitcoin.icns
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents
/usr/bin/install -c -m 644 share/qt/Info.plist Bitcoin-Qt.app/Contents/Info.plist
Making all in src
Making all in doc/man
make[1]: Nothing to be done for `all'.
make[1]: Nothing to be done for `all-am'.
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents/MacOS
STRIPPROG="/usr/bin/strip" /bin/sh /Users/doge/bitcoin/build-aux/install-sh -c -s  ./src/qt/bitcoin-qt Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents/Resources/Base.lproj
echo '{	CFBundleDisplayName = "Bitcoin Core"; CFBundleName = "Bitcoin Core"; }' > Bitcoin-Qt.app/Contents/Resources/Base.lproj/InfoPlist.strings
/opt/homebrew/bin/python3.9 ./contrib/macdeploy/macdeployqtplus Bitcoin-Qt.app Bitcoin-Core -translations-dir=/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/translations -dmg
Traceback (most recent call last):
  File "/Users/doge/bitcoin/./contrib/macdeploy/macdeployqtplus", line 22, in <module>
    from ds_store import DSStore
  File "/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/native/lib/python3/dist-packages/ds_store/__init__.py", line 1, in <module>
    from .store import DSStore, DSStoreEntry
  File "/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/native/lib/python3/dist-packages/ds_store/store.py", line 8, in <module>
    import mac_alias
  File "/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/native/lib/python3/dist-packages/mac_alias/__init__.py", line 1, in <module>
    from .alias import *
  File "/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/native/lib/python3/dist-packages/mac_alias/alias.py", line 16, in <module>
    from . import osx
  File "/Users/doge/bitcoin/depends/arm-apple-darwin20.2.0/native/lib/python3/dist-packages/mac_alias/osx.py", line 424, in <module>
    _statfs = libc['statfs$INODE64']
  File "/opt/homebrew/Cellar/python@3.9/3.9.1_4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(0x102b9d810, statfs$INODE64): symbol not found
make: *** [Bitcoin-Core.dmg] Error 1
```

**Comment by jonasschnelli** [comment]
@awesome-doge 
`make deploy` currently doesn't work on Apple M1 (ARM64) due to missing support in the dmg tools area (ds_store AFAIK).
I added it to the PR descriptions list (as open task).

**Comment by awesome-doge** [comment]
It seems that someone opened a related issue

build: `make deploy` fails on macOS Big Sur 11.1 #20878

https://github.com/bitcoin/bitcoin/issues/20878

**Comment by sgulls** [comment]
[Changing](https://github.com/bitcoin/bitcoin/pull/21555) mac_alias to [2.2.0](https://github.com/al45tair/mac_alias/issues/10) fixes `make deploy`

**Comment by sgulls** [comment]
@jonasschnelli `make deploy` fix has been [committed](https://github.com/bitcoin/bitcoin/commit/2cbb4ab7a92a281d7d7b4821654515dd59383ff3) to `master`

**Comment by sgulls** [comment]
@michaelsblank yes, that's what "committed to master" [means](https://www.atlassian.com/git/tutorials/saving-changes/git-commit)

**Comment by michaelsblank** [comment]
> @michaelsblank yes, that's what "committed to master" [means](https://www.atlassian.com/git/tutorials/saving-changes/git-commit)

Yes, yes I know. I'm a developer. But, I should have been more precise. Does this now mean we can use the normal MacOS readme guide but for this latest commit and everything should build and run? I just found it odd that when I previewed the commit the change was just a version number and a hash change. If that got it running awesome, I just didn't know if there was more work to be done. Excuse me for my poor wording before.

**Comment by sgulls** [comment]
@michaelsblank 

I [mentioned](https://github.com/bitcoin/bitcoin/issues/20371#issuecomment-810642505) the pertinent [issue](https://github.com/al45tair/mac_alias/issues/10) with one of the packages and in the issue, the maintainer said they fixed the issue that was encountered when building the app bundle and dmg. That fix was included in their [version bump](https://github.com/al45tair/mac_alias/issues/10#issuecomment-741145082).

**Comment by jsarenik** [comment]
@michaelsblank You seem to be one of few who have M1 hardware. Hands-on experience is welcome so feel free to write more here. Thanks!

**Comment by ipa-tys** [comment]
Hi, I also have a Mac Mini M1 (up-to-date with Big Sur 11.3.1) and was able to previously compile 0.21.0. But now, when trying to update to 0.21.1, I ran into an error when compiling the dependencies by running 'make' in /depends subdirectory:

`ERROR: Feature 'system-zlib' was enabled, but the pre-condition 'libs.zlib' failed. make: *** [/Users/xxx/bitcoin/depends/work/build/arm-apple-darwin20.2.0/qt/5.9.8-82a1b93ac0c/qtbase/.stamp_configured] Error 3`

I thought maybe zlib was not installed, so I ran 'brew install zlib', producing this output:

`==> Pouring zlib--1.2.11.arm64_big_sur.bottle.tar.gz
==> Caveats
zlib is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

For compilers to find zlib you may need to set:
  export LDFLAGS="-L/opt/homebrew/opt/zlib/lib"
  export CPPFLAGS="-I/opt/homebrew/opt/zlib/include"

For pkg-config to find zlib you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/zlib/lib/pkgconfig"
`

Even when adding these environment variables, I got the same error as shown above when trying to compile the dependencies, so I guess that attempt did not help. 

Any advice?

**Comment by ipa-tys** [comment]
Since it's been 10 days, I hope it's ok to give this a small bump, @jonasschnelli.
As written above, I'd be happy to help testing on M1 architecture, but from the lack of feedback it seems there is currently not so much interest in it?

**Comment by fanquake** [comment]
@ipa-tys building depends on an M1 machine should work fine at the moment. I know there are multiple people successfully building that way. If you are still having problems, then please open an issue with all the relevant info/logs. Make sure you've tried building depends from scratch etc.

If you're interested in cross-compilation for M1, you can check out #21851.

**Comment by ipa-tys** [comment]
Thanks! Yes I tried building depends from scratch (which, as I mentioned above, had worked for me in 0.21.0). I will file a separate issue, as you suggested.

**Comment by sgulls** [comment]
@jonasschnelli 
I've managed to cross compile from x86_64-linux-gnu to aarch64-apple-darwin20.4. I built [osxcross](https://github.com/tpoechtrager/osxcross) against the latest stable SDK (SDK version 11.3 from Xcode 12.5), added osxcross binaries to `$PATH` built the depends folder with 
`make HOST=aarch64-apple-darwin20.4 FORCE_USE_SYSTEM_CLANG=1 XCODE_VERSION=12.5 XCODE_BUILD_ID=12E262`
and successfully compiled a working GUI app bundle that ran on my arm64 Mac

**Comment by prusnak** [comment]
> Support for AArch64 NI SHA2 instructions should/could also be added.

Added in https://github.com/bitcoin/bitcoin/pull/24115

**Comment by sipa** [comment]
Does Apple M1 have the "rng" extension (true RNG built into the CPU)?

**Comment by prusnak** [comment]
Since https://github.com/bitcoin/bitcoin/pull/21851 and others are merged already and 23.0 will have native arm64-darwin build I think we can close this issue?

**Comment by prusnak** [comment]
> Does Apple M1 have the "rng" extension (true RNG built into the CPU)?

@sipa It seems it does - https://support.apple.com/en-gb/guide/security/sec59b0b31ff/web (search for "True Random Number Generator")
